### DISCUSSION TITLE
is it possible to add an audit exception for brew test-bot ie. gha for flat namespaces

### DISCUSSION THREAD

**Discussion by ipatch** [Homebrew/discussions#2591] [Tap maintenance and Homebrew development] [answered]
i'm getting closer to being able to build a `FreeCAD.app` bundle using macos homebrew for amd64, however with the most recent run i came across an issue where the `brew audit` from the test-bot (i presuming its _test-bot_) that threw error due to open-mpi being compiled with _flat namespaces_.

i'm not familiar with open-mpi or namespaces other than the issues i've read on github and following the breadcrumbs.

there seems to be an open issue with the open-mpi project / github repo,

https://github.com/open-mpi/ompi/issues/7766

that discusses this very topic i am bringing up. also i noticed within that topic a member that i see quite frequently in these discussions linked to it.

i'd rather not try and additional versioned formula of open-mpi into the freecad tap, but rather would have some sort "audit exception" if possible that will allow the gha / test-bot to complete it's process without throwing an error.

1. is it possible that i can use open-mpi with freecad and still have it's libraries use flat namepsaces?
2. add a "audit exception" to the tap or something of that nature that allows the ci/cd to complete without throwing an error?
3. where do i locate a list of exceptions supported by homebrew, test-bot that applicable for homebrew taps?

**Comment by Bo98** [comment, marked as answer]
Yes. We in fact do this for our own open-mpi: https://github.com/Homebrew/homebrew-core/blob/e4a0631c29b0f55a4870cc8da399419c4e587202/audit_exceptions/flat_namespace_allowlist.json#L11

We should probably document `audit_exceptions` somewhere, but for now check what homebrew-core does.

**Reply by ipatch** [reply]
thank you.

i did a couple of searches for `audit` and didn't get any results

https://www.rubydoc.info/github/Homebrew/brew

**Reply by ipatch** [reply]
also, i assume taps that use gha are able to use these same audit exceptions for 3rd party taps, no?

**Reply by Bo98** [reply]
Yes, with any CI.

**Reply by ipatch** [reply]
okay little followup i added that audit exception list to the homebrew freecad tap however since those formula files are contained within a different tap ie. homebrew-core how would i tell the homebrew-freecad tap that (our tap) homebrew-freecad tap needs to add those exceptions for homebrew-core?

basically when i'm running, ie. **example here**

```
brew audit freecad/freecad/pivy@0.6.6 --online --git --skip-style   
```

i'm getting the below output

```
freecad/freecad:
  * audit_exceptions/flat_namespace_allowlist.json references
    formulae or casks that are not found in the freecad/freecad tap.
    Invalid formulae or casks: homebrew/homebrew-core/mpich, homebrew/homebrew-core/open-mpi
Error: 1 problem in 1 tap detected
```

**Reply by ipatch** [reply]
didn't get great results using the below search on github

https://github.com/search?p=15&q=path:/audit_exceptions&type=Code

**Reply by SMillerDev** [reply]
It looks like your homebrew-core tap is just outdated since this is already defined for our open-mpi

**Reply by ipatch** [reply]
hmmm, no it's a matter of the homebrew-freecad tap ci/cd (possibly brew test-bot) finding/noticing that open-mpi is compiled with flat namepsace. so when i attempt to bottle freecad for the homebrew-freecad our ci is catching this exception.

https://github.com/FreeCAD/homebrew-freecad/runs/4412564484?check_suite_focus=true#step:13:30

**Reply by ipatch** [reply]
it looks like the ruby file that peforms this check takes a "name". Can the name not contain a tap prefix of some sort? ...i guess the "name" would have be setup to have different variances of what a "name" is... 🤔

![image](https://user-images.githubusercontent.com/613805/144762677-0ec59d19-26b5-4d11-8a84-54c437c2d02c.png)

**Reply by Bo98** [reply]
Yes individual taps are responsible for their own formulae - not other tap's formulae. Equally, you won't ever see an audit failure for another tap's formula unless you are running test-bot on it for some unusal reason.

Looking at your logs, you want to add the exception for `freecad`. It is the freecad formula which it is complaining about - not open-mpi (because it is copying the open-mpi dylibs into the freecad install rather than just depending on open-mpi).

```
freecad/freecad/freecad:
    * Libraries were compiled with a flat namespace.
      This can cause linker errors due to name collisions, and
      is often due to a bug in detecting the macOS version.
        /usr/local/Cellar/freecad/0.19/FreeCAD.app/Contents/lib/libmpi.40.dylib
        /usr/local/Cellar/freecad/0.19/FreeCAD.app/Contents/lib/libopen-orted-mpir.dylib
        /usr/local/Cellar/freecad/0.19/FreeCAD.app/Contents/lib/libopen-pal.40.dylib
        /usr/local/Cellar/freecad/0.19/FreeCAD.app/Contents/lib/libopen-rte.40.dylib
```

**Reply by ipatch** [reply]
wow thanks for the clarification on that.
